docs/troubleshooting.md

```diff
@@ -0,0 +1,172 @@
+
+### Solution
+
+```js
+router.get('/fetch', function (ctx, next) {
+  return models.Book.findById(parseInt(ctx.query.id)).then(function (book) {
+    ctx.body = book;
+  });
+});
+```
+
+Returning the promise given by the database interface tells Koa to wait for the promise to finish before responding. At that time, the body will have been set. This results in Koa sending back a 200 with a proper response.
+
+The fix in the `async` version is to add an `await` statement:
+
+```js
+router.get('/fetch', async (ctx, next) => {
+  await models.Book.findById(parseInt(ctx.query.id)).then(function (book) {
+    ctx.body = book;
+  });
+});
+```
+
+## My middleware is not called
+
+This can be due to an interrupted chain of middleware calls.  This can cause a 404 if the
+middleware that is skipped is responsible for the `ctx.body` or `ctx.status` setter.
+This is less common than [My response or context changes have no effect](#my-response-or-context-changes-have-no-effect),
+but it can be a much bigger pain to troubleshoot.
+
+### Problematic code
+
+```js
+router.use(function (ctx, next) {
+  // Don't Repeat Yourself! Let's parse the ID here for all our middleware
+  if (ctx.query.id) {
+    ctx.state.id = parseInt(ctx.query.id);
+  }
+});
+
+router.get('/fetch', function (ctx, next) {
+  return models.Book.findById(ctx.state.id).then(function (book) {
+    ctx.body = book;
+  });
+});
+```
+
+### Cause
+
+In the code above, the book is never fetched from the database, and in fact our route was never called. Look closely at our helper middleware. We forgot to `return next()`! This causes the middleware flow to never reach our route, ending our "helper" middleware.
+
+### Identifying this as the issue
+
+Identifying this problem is easier than most, add a log at the beginning of the route. If it doesn't trigger, your route was never reached in the middleware chain.
+
+```js
+router.use(function (ctx, next) {
+  // Don't Repeat Yourself! Let's parse the ID here for all our middleware
+  if (ctx.query.id) {
+    ctx.state.id = parseInt(ctx.query.id);
+  }
+});
+
+router.get('/fetch', function (ctx, next) {
+  console.log('Route called'); // Never happens
+  return models.Book.findById(ctx.state.id).then(function (book) {
+    ctx.body = book;
+  });
+});
+```
+
+To find the middleware causing the problem, try adding logging at various points in the middleware chain.
+
+### Solution
+
+The solution for this is rather easy, simply add `return next()` to the end of your helper middleware.
+
+```js
+router.use(function (ctx, next) {
+  if (ctx.query.id) {
+    ctx.state.id = parseInt(ctx.query.id);
+  }
+  return next();
```

Comment:
does the `return` actually have significance? I forget what we did with Koa 2.x now haha..